### ISSUE TITLE
editor:select-to-end-of-line should work similarly to editor:move-to-end-of-line

### ISSUE THREAD

**Issue by alexandernst** [atom/atom#7159]
I'd normally use the `Shift + End` buttons to select the text from the current cursor position to the end of the line, but this doesn't seem to be working with Atom.

Let's say I have this long line:

```
This is a very long line of text that is used to test a bug report in Atom.
```

Now let's say I resize Atom's window so the line can't fit in a single row, specifically, the text `port in Atom.` is forced on a new row.

Now,  I'd expect that if I place the cursor at the beginning of the line and hit `Shift + End`, the entire text is selected. I'd even accept as a reasonable behavior if Atom selects everything up to the `test a bug re` part, but that is not the case. Atom is actually selecting everything on the first row of the line, except the last character. That is, in my example, after placing the cursor at the beginning of the line and hitting `Shift + End`, Atom is selecting this:

```
This is a very long line of text that is used to test a bug r
```

**Comment by izuzak** [comment]
@alexandernst I'm having trouble reproducing this, so can you include [more information](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#submitting-issues) for your report, and also check if you can reproduce this in [safe mode](https://atom.io/docs/latest/hacking-atom-debugging#check-if-the-problem-shows-up-in-safe-mode)?

One thing that's odd is that you say:

> Now let's say I resize Atom's window so the line can't fit in a single row, specifically, the text `port in Atom`. is forced on a new row.

But as far as I know, Atom will not soft wrap text into a new line by breaking a word in two. So, the whole word `report` should be wrapped to another line. Can you post a screenshot where Atom is breaking the word in half when soft-wrapping? Also, can you share a GIF which demonstrates this behavior in safe mode (incorrect selecting at the end of the screen line), with the [keybinding resolver](https://atom.io/docs/latest/hacking-atom-debugging#check-the-keybindings) visible?

**Comment by alexandernst** [comment]
@izuzak Oh, indeed, you're right. The steps I gave to reproduce the problem are incorrect. Let me try again:

1. Disable soft wrap
2. Write a very long text without any spaces
3. Shrink the window so Atom will be forced to split part of the string to a new row
4. Do the shift + end thing

Video: https://github.com/atom/atom/issues/7159

**Comment by izuzak** [comment]
@alexandernst If soft wrap is disabled (your step 1), then Atom will not split the string into a new row (your step 3). Can you clarify those instructions one more time? Also, did you try to put a video at the end of that comment? I'm not seeing it -- I see only a link back to this issue. Thanks!

**Comment by alexandernst** [comment]
@izuzak Actually Atom is splitting the text even if I have `Soft wra` disabled. Please check my video. It's very detailed, recorded in HQ and it contains a screenshot of my settings.

**Comment by izuzak** [comment]
> Please check my video. It's very detailed, recorded in HQ and it contains a screenshot of my settings.

There is no video in your previous comment, @alexandernst -- it's just a link back to this issue:

![screen shot 2015-06-08 at 16 11 20](https://cloud.githubusercontent.com/assets/38924/8036582/07e33cf4-0df9-11e5-9b81-be0857f8b0c5.png)

**Comment by alexandernst** [comment]
@izuzak Oops... ctrl + v is playing me tricks today. Here you go: https://www.youtube.com/watch?v=oSseAHdX1TY

**Comment by izuzak** [comment]
Thanks, @alexandernst -- that's very helpful, I was able to reproduce that behavior now. (And you do have soft wrap turned on in that editor. Try running the "Editor: toggle soft wrap" command from the command palette and see how wrapping changes. What you see in the settings view is the default for new editors, I believe -- it doesn't switch the soft wrap for existing editors.)

@nathansobo I think this is another one for you to look at. :basketball:

**Comment by kylesimmonds** [comment]
Is there any work being done on this? I'm trying to select an entire line that is soft wrapped, but it only selects the screen line it's on.

In the Keybindings list, the command is `editor:select-to-end-of-line`, but it only selects to the end of the screen line.

**Comment by davidspiess** [comment]
@kylesimmonds  try `editor:move-to-end-of-line`. I think it does was you expect.

**Comment by palsivertsen** [comment]
I see a similar behavior when using the `end` key. This is bound to `editor:move-to-end-of-screen-line` by default. Adding a custom keybinding as follows changes the the behavior as desired:
`'atom-text-editor':`
`  'end': 'editor:move-to-end-of-line'`

The `shift-end` keybinding is bound to `editor:select-to-end-of-line`. Looking at the command names I'd expect `editor:move-to-end-of-line` and `editor:select-to-end-of-line` to behave the same way, but it does not.

### Suggested fix
- Rename `editor:select-to-end-of-line` to `editor:select-to-end-of-screen-line`.
- Implement new command `editor:select-to-end-of-line` with similar behavior as `editor:move-to-end-of-line`
- (Optional) Assign default keybinding `shift-end` to `editor:select-to-end-of-screen-line` to keep current default behavior.

**Comment by lee-dohm** [comment]
Thanks @palsivertsen for the great summary! You are correct, I would agree that `move-to-end-of-line` and `select-to-end-of-line` should act similarly. Renamed the issue to be more descriptive.

I also tested this on ST3, TextMate 2 and OS X's TextEdit. <kbd>Shift+End</kbd> behaved the way that was described in all of those applications.

**Comment by gw0** [comment]
Maybe behavior could be similar to Sublime Text where one press of `end` (or `shift+end`) moves/selects till the end of screen line, but if you press it again it moves/selects till the actual line.

**Comment by kylesimmonds** [comment]
I just made a pull request to address number 1 in the comment above: https://github.com/atom/atom/issues/7159#issuecomment-195021280

Side note, this was my first pull request ever, so any tips for improvement are appreciated :grin:

**Comment by alexandernst** [comment]
I think this is still a valid issue @izuzak

**Comment by ManuelTS** [comment]
I'm on windows 10 with atom 1.24.0 x64 with _soft wrap_ activated, when I press _Shift + End_ the selection goes from the curser to the line break of the soft wrap but not from the curser to the end of the line. Here a picture: ![untitled](https://user-images.githubusercontent.com/8605616/36416116-6778f7b4-1628-11e8-906f-23924c23c1ab.png)

**Comment by ariasuni** [comment]
This is really annoying when using multiple cursors with long lines, because depending on the size of the line, the cursor will not always go to the end of the real line.

A partial workaround is to use:

```cson
'atom-text-editor':
    'alt-home': 'editor:move-to-beginning-of-line'
    'alt-end': 'editor:move-to-end-of-line'
```

But I can’t use shift in combination with those, which is unintuitive.

**Comment by ariasuni** [comment]
By the way, I’m using the config below now, I think it’s self-descriptive:

```cson
'atom-text-editor':
    'alt-home': 'editor:move-to-beginning-of-line'
    'alt-end': 'editor:move-to-end-of-line'
    'alt-shift-home': 'editor:select-to-beginning-of-line'
    'alt-shift-end': 'editor:select-to-end-of-line'
```

**Comment by MPvHarmelen** [comment]
## version of Atom
1.53.0

## OS and version
Ubuntu 20.04 LTS 64-bit

## Steps to reproduce
1. Add the following (thanks @ariasuni) to the start of your keymap file:
   ```cson
   'atom-text-editor':
       'alt-home': 'editor:move-to-beginning-of-line'
       'alt-end': 'editor:move-to-end-of-line'
       'alt-shift-home': 'editor:select-to-beginning-of-line'
       'alt-shift-end': 'editor:select-to-end-of-line'
   ```
2. open a new tab
3. create a very long line (e.g. by typing `lorem` and then hitting <kbd>tab</kbd>)
4. enable soft wrap (<kbd>ctrl-shift-p</kbd> -> "toggle soft-wrap")
5. go to the start of the long line (<kbd>alt-home</kbd>)
6. hit <kbd>alt-shift-end</kbd> and notice how only the first soft line is selected, even though hitting <kbd>alt-end</kbd> move the cursor to the end of the actual line, and <kbd>alt-home</kbd> selects the whole actual line.
